qrexec/policy/parser.py

```diff
@@ -0,0 +1,1252 @@
+
+        Raises:
+            PolicySyntaxError: when syntax error is found
+        '''
+
+        try:
+            service, argument, source, target, action, *params = line.split()
+        except ValueError:
+            raise PolicySyntaxError(filepath, lineno, 'wrong number of fields')
+
+        return cls(service, argument, source, target, action, params,
+            policy=policy, filepath=filepath, lineno=lineno)
+
+    @classmethod
+    def from_line_service(cls, policy, service, argument, line, *,
+            filepath, lineno):
+
+        try:
+            source, target, action, *params = line.split()
+        except ValueError:
+            raise PolicySyntaxError(filepath, lineno, 'wrong number of fields')
+
+        params = tuple(itertools.chain(*(p.split(',') for p in params)))
+
+        return cls(service, argument, source, target, action, params,
+            policy=policy, filepath=filepath, lineno=lineno)
+
+    def is_match(self, request):
+        '''Check if given (service, argument source, target) matches this line.
+
+        :param system_info: information about the system - available VMs,
+            their types, labels, tags etc. as returned by
+            :py:func:`app_to_system_info`
+        :param service: name of the service
+        :param argument: the argument
+        :param source: name of the source VM
+        :param target: name of the target VM, or None if not specified
+        :param system_info: the context
+        :return: True or False
+        '''
+
+        return (self.is_match_but_target(request)
+            and self.target.match(request.target, source=request.source,
+                system_info=request.system_info))
+
+    def is_match_but_target(self, request):
+        '''Check if given (service, argument source) matches this line.
+
+        Target is ignored. This is used for :py:meth:`collect_targets_for_ask`.
+
+        :param system_info: information about the system - available VMs,
+            their types, labels, tags etc. as returned by
+            :py:func:`app_to_system_info`
+        :param service: name of the service
+        :param argument: the argument
+        :param source: name of the source VM
+        :param target: name of the target VM, or None if not specified
+        :param system_info: the context
+        :return: True or False
+        '''
+
+        return ((self.service is None or self.service == request.service)
+            and (self.argument is None or self.argument == request.argument)
+            and self.source.match(request.source,
+                    system_info=request.system_info))
+
+
+class AbstractPolicyParser(metaclass=abc.ABCMeta):
+    '''A minimal, pluggable, validating policy parser'''
+
+    #: default rule type
+    rule_type = Rule
+
+    @staticmethod
+    def resolve_file_and_filepath(file, filepath):
+        if filepath is None:
+            if isinstance(file, (str, pathlib.Path)):
```

Comment:
This function has two purposes:

1. API flexibility, or as you called it, "confusing file and filepath". This is common in many Python external-facing functions that when you are supposed to supply "file", you can either pass its path or `open()`'d file. This can be either path to the file (`str` or `Path`, the choice is called "[path-like](https://docs.python.org/3/glossary.html#term-path-like-object)" in Python >= 3.6), or file-like object (from `open()` or `io.StringIO()`, or in the future even [`TarFile.extractfile()`](https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractfile) for backups). For the file case, it is possible to write fake name (convenient in tests) or true name that is not available, because it is in TarInfo.

2. Hook point for overloading, if you'd like to resolve paths differently. This is used in TestPolicy, used for tests to load policy from memory: https://github.com/QubesOS/qubes-core-qrexec/pull/2/files#diff-25b0a4971b66e6b4d6eca5fd9bb60ddcR1139. The intention is to be able to support `!include` in tests.

You're right that docstring is missing.

I couldn't come up with qualitatively better design, but one alternative is to have `load_policy_file[_service]()` to always accept file-like object and call this function from `handle_*` methods. I think it is not a big difference from audit POV, so I'd reap the benefit of more flexible API.

Reply:
Ok, I'm convinced. But there are few issues: if `file` is path-like (or None...), but `filepath` is not None, this will not work. I don't think I've found such use in the code, but this is external-facing API and it should be handled (or prohibited explicitly - with a docstring and maybe even assert).
Then, you do allow this function to return `filepath=None` (the `AttributeError` case). Make sure all the consumers of `filepath` do accept it. Or error out when `filepath` cannot be determined.

Reply:
There are two uses for `filepath`: error reporting and path traversal (`!include` and friends). For the first case `filepath` can be `None` and this is handled in [`PolicySyntaxError`](https://github.com/QubesOS/qubes-core-qrexec/pull/2/files#diff-5a799fa121d4d5e6e5e60950ff5a0026R30). But this only happens in synthetic cases, like instantiating objects manually and in tests. The second use of filepath is much more serious, but 1) `None` won't happen in practice, because for the common case when policy is loaded from files `filepath` is not `None` 2) if it happened anyhow, it will barf TypeError on `/` operations.

Tomorrow I'll experiment with the other approach. I didn't really try it, so maybe it will be easier to review after all. Your two comments are counterargument to my hypothesis that it is easy to read.